metagraph/src/annotation/taxonomy/taxonomic_db.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef __TAXONOMIC_DB_HPP__
+#define __TAXONOMIC_DB_HPP__
+
+#include <tsl/hopscotch_set.h>
+#include <tsl/hopscotch_map.h>
+
+#include "annotation/representation/base/annotation.hpp"
+#include "graph/annotated_dbg.hpp"
+
+
+namespace mtg {
+namespace annot {
+
+/**
+ * TaxonomyDB constructs a taxonomic map (kmer to taxid) in a similar way
+ * to how Kraken2 works. The file exported by this object will be further
+ * used for taxonomic sequence classifier in './metagraph tax_class'.
+ */
+class TaxonomyDB {
+  public:
+    typedef std::string AccessionVersion;
+    typedef std::uint64_t TaxId;
+    typedef std::uint64_t NormalizedTaxId;
+    typedef std::vector<std::vector<NormalizedTaxId>> ChildrenList;
+    typedef annot::MultiLabelEncoded<AccessionVersion> Annotator;
+    using KmerId = Annotator::Index;
+
+  private:
+    /**
+     * node_depth returns the depth for each node.
+     * The root is the unique node with maximal depth and all the leaves have depth 1.
+     */
+    std::vector<uint64_t> node_depth;
+
+    /**
+     * rmq_data[0] contains the taxonomic tree linearization
+     *          (e.g. for root 1 and edges={1-2; 1-3}, the linearization is "1 2 1 3 1")
+     * rmq_data[l][x] returns the node with the maximal depth among positions [x, x+2^l-1]
+     * in the linearization.
+     *          (e.g. rmq_data[3][6] return the node with max depth in [6, 13])
+     */
+    std::vector<std::vector<NormalizedTaxId>> rmq_data;
+
+    /**
+     * node_to_linearization_idx[node] returns the index of the first occurrence of node
+     * in the linearization.
+     */
+    std::vector<uint64_t> node_to_linearization_idx;
+
+    /**
+     * lookup_table maps accession version to taxid.
+     */
+    tsl::hopscotch_map<AccessionVersion, TaxId> lookup_table;
+
+    /**
+     * precalc_log2 is a table for a fast compute of log2(x).
+     */
+    std::vector<uint64_t> precalc_log2;
+
+    /**
+     * precalc_pow2 is a table for a fast compute of pow2(x).
+     */
+    std::vector<uint64_t> precalc_pow2;
+
+    /**
+     * Maps taxid to its normalized index. Used for optimizing the runtime performance.
+     */
+    tsl::hopscotch_map<TaxId, NormalizedTaxId> normalized_taxid;
+
+    /**
+     * Maps normalized taxid to its denormalized counterpart.
+     */
+    std::vector<TaxId> denormalized_taxid;
+
+    /**
+    * taxonomic_map[kmer] returns the taxid LCA for the given kmer.
+    */
+    sdsl::int_vector<> taxonomic_map;
+
+    /**
+     * Reads and returns the taxonomic tree
+     *
+     * @param [input] taxo_tree_filepath path to a "nodes.dmp" file.
+     * @param [output] tree -> tree stored as list of children.
+     * @param [output] root_node -> normalized id of the root of the tree.
+     */
+    void read_tree(const std::string &taxo_tree_filepath,
+                   ChildrenList &tree, NormalizedTaxId &root_node);
```

Comment:
```suggestion
    void read_tree(const std::string &taxo_tree_filepath,
                   ChildrenList *tree,
                   NormalizedTaxId *root_node);
```

Reply:
Each parameter should be on a different line?

Reply:
Yes, if they don't fit in one line.
But not necessarily when you call a function.